### PR TITLE
drivers: sd_card: Add SD card read/write support

### PR DESCRIPTION
**Pull request by mchindri**
This interface supports byte read and write operations for SD cards
that meet the following conditions:
*   	Version 2.00 or later
*   	High capacity or extended capacity (SDHX or SDXC)
*   	Supply voltage of 3.3V

Signed-off-by: Mihail Chindris <mihail.chindris@analog.com>

### CODE REVIEW

**Comment by mchindri** [comment]
V2
 - Fix astyle

File: drivers/sd_card/sd.c [hunk 1]
```diff
@@ -0,0 +1,568 @@
...
+* MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
+* IN NO EVENT SHALL ANALOG DEVICES BE LIABLE FOR ANY DIRECT, INDIRECT,
+* INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+* LIMITED TO, INTELLECTUAL PROPERTY RIGHTS, PROCUREMENT OF SUBSTITUTE GOODS OR
+* SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+* CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+* OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+* OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*******************************************************************************/
+
```
Comment (amiclaus) [inline drivers/sd_card/sd.c]: add please 
https://github.com/analogdevicesinc/no-OS/blob/6d914f12acfcdc631bcf41684fbc3846d747f971/drivers/platform/altera/spi.c#L40-L42

File: drivers/sd_card/sd.c [hunk 2]
```diff
@@ -0,0 +1,568 @@
...
+#include "sd.h"
+
```
Comment (amiclaus) [inline drivers/sd_card/sd.c]: add please:
https://github.com/analogdevicesinc/no-OS/blob/6d914f12acfcdc631bcf41684fbc3846d747f971/drivers/platform/altera/spi.c#L51-L53

Comment (mchindri) [reply drivers/sd_card/sd.c]: shouldn't it be Functions Implementation in the .c files?

File: drivers/sd_card/sd.h [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+#include <stdlib.h>
+#include "spi.h"
+#include "error.h"
+
+/******************************************************************************/
+/********************** Macros and Constants Definitions **********************/
+/******************************************************************************/
+
+
+#define BIT_CCS					(1u<<30)
```
[the remark this answers is not in the document]
Comment (mchindri) [reply drivers/sd_card/sd.h]: I think macros includes should be in the header file like here:
https://github.com/analogdevicesinc/no-OS/blob/master/drivers/accel/adxl345/adxl345.h

File: drivers/sd_card/sd.c (hunk 2 above)
Comment (amiclaus) [reply drivers/sd_card/sd.c]: `*.h` -> Functions Declarations
`*.c` -> Functions Definitions

File: drivers/sd_card/sd.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (scuciurean) [reply drivers/sd_card/sd.h]: That ones are mostly register addresses and chip commands, needed by the user when using the driver.
If the macro/definition isn't used by the user, it should not be shared.

File: drivers/sd_card/sd.c [hunk 4]
```diff
@@ -0,0 +1,568 @@
...
+ * @param sd_desc - Instance of the SD card
+ * @return SUCCESS in case of success, FAILURE otherwise.
+ */
+static uint32_t wait_until_not_busy(struct sd_desc *sd_desc)
+{
+	uint8_t	data;
+
+	data = 0x00;
+	while (data == 0x00) {
+		data = 0xFF;
```
[the remark this answers is not in the document]
Comment (mchindri) [reply drivers/sd_card/sd.c]: In the SD card specification says that the card always sends a response.
Should I add a timeout considering this?

File: drivers/sd_card/sd.c [hunk 5]
```diff
@@ -0,0 +1,568 @@
...
+
+#include "sd.h"
+
+/**
+ * Read SD card bytes until one is different from 0xFF
+ * @param sd_desc	- Instance of the SD card
+ * @param data_out	- The read bytes is wrote here
+ * @return SUCCESS in case of success, FAILURE otherwise.
+ */
+static uint32_t wait_for_response(struct sd_desc *sd_desc, uint8_t *data_out)
```
Comment (adrimbarean) [inline drivers/sd_card/sd.c]: `FAILURE` defined in the `error.h` header has value `-1`, but this function cannot return a signed integer.

**Comment by dbogdan** [comment]
For consistency with the rest of the folders, please rename the sd_card to sd-card.

**Comment by mchindri** [comment]
V3
 Add comments of sections
 Update description of sd_desc
 Move macros to .c file
 Update return type from uint32_t to int32_t
 Reaname `sd_card` -> `sd-card`

File: drivers/sd-card/sd.c [hunk 6]
```diff
@@ -0,0 +1,614 @@
...
+						  cmd_desc->response_len - 1))
+			return FAILURE;
+	}
+
+	return SUCCESS;
+}
+
+/**
+ * Send one block of data to the SD card
+ * @param sd_desc		- Instance of the SD card
```
Comment (dbogdan) [inline drivers/sd-card/sd.c]: Make sure your editor is configured to have tabs of 8 characters.

Comment (dbogdan) [reply drivers/sd-card/sd.c]: The astyle check is also failing.

**Comment by mchindri** [comment]
V3
 - Fix astyle
 - Fix tabs in comments

**Comment by mchindri** [comment]
V3
 - Fix astyle
 - Fix tabs in comments

File: drivers/sd-card/sd.c [hunk 7]
```diff
@@ -0,0 +1,614 @@
...
+ * @param len		- Length of data in bytes
+ * @return SUCCESS in case of success, FAILURE otherwise.
+ */
+int32_t sd_write(struct sd_desc *sd_desc, uint8_t *data, uint64_t address,
+			uint64_t len)
+{
+	struct cmd_desc	cmd_desc;
+	uint32_t	nb_of_blocks;
+	uint32_t	first_block_addr;
+	uint8_t		first_block[DATA_BLOCK_LEN];
```
Comment (adrimbarean) [inline drivers/sd-card/sd.c]: Might need to be aligned to 4 bytes.

**Comment by mchindri** [comment]
V4
 - Aling static buffers for posible microcontrolers that need 4 bytes alignement
 - Add timout with counter in  `wait_for_response `and `wait_until_not_busy` until implementation of timer.c for ADUCM3029
 - Added function `get_nb_of_blocks` to calculate the number of blocks instead of passing it as paramenter

**Comment by mchindri** [comment]
v5
 - Fix asyle

**Comment by mchindri** [comment]
V5 bis
 - Astyle for sd.h

**Comment by mchindri** [comment]
V6
 - Add in comment for implementation of timeout the `TODO` keyword

**Comment by mchindri** [comment]
Will merge today if no more comments

**Pull request merged**